### ISSUE TITLE
ambiguous operator *= with hipFloatComplex

### ISSUE THREAD

**Issue by mgates3** [ROCm-Developer-Tools/HIP#2797]
Tested with ROCm 5.2.
```
> hipcc --version
HIP version: 5.2.21151-afdc89f8
AMD clang version 14.0.0 (https://github.com/RadeonOpenCompute/llvm-project roc-5.2.0 22204 50d6d5d5b608d2abd6af44314abc6ad20036af3b)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /opt/rocm-5.2.0/llvm/bin
```

Reproducer:
```
// File complex.hip.cc
#include <hip/hip_runtime.h>
#include <hip/hip_complex.h>

__global__
void kernel( hipFloatComplex x, hipFloatComplex y )
{
    // This next line produces a compile error,
    // because hipFloatComplex is simply a typedef of HIP_vector_type.
    x *= y;

    // This works.
    x = x * y;
}

void driver( hipFloatComplex x, hipFloatComplex y, hipStream_t stream )
{
    hipLaunchKernelGGL( kernel, dim3(1), dim3(1), 0, stream, x, y );
}

int main( int argc, char** argv )
{
    hipFloatComplex x, y;
    hipStream_t stream;
    driver( x, y, stream );
    return 0;
}
```

Compile error:
```
> hipcc -o complex complex.hip.cc 
complex.hip.cc:9:7: error: use of overloaded operator '*=' is ambiguous (with operand types 'hipFloatComplex' (aka 'HIP_vector_type<float, 2>') and 'hipFloatComplex')
    x *= y;
    ~ ^  ~
/opt/rocm/include/hip/amd_detail/amd_hip_vector_types.h:838:26: note: candidate function
        HIP_vector_type& operator*=(const HIP_vector_type& x) noexcept
                         ^
/opt/rocm/include/hip/amd_detail/amd_hip_complex.h:236:1: note: candidate function
COMPLEX_MUL_PREOP_OVERLOAD(hipFloatComplex)
^
/opt/rocm/include/hip/amd_detail/amd_hip_complex.h:110:41: note: expanded from macro 'COMPLEX_MUL_PREOP_OVERLOAD'
    __HOST_DEVICE__ static inline type& operator*=(type& lhs, const type& rhs) {               \
                                        ^
1 error generated when compiling for gfx90a.
```

**Comment by ppanchad-amd** [comment]
@mgates3 Apologies for the lack of response. Can you please test with latest ROCm 6.0.2 (HIP 6.0.32831)? If resolved, please close ticket. Thanks!

**Comment by mgates3** [comment]
The issue is still in 6.0.2:
```
sh dopamine hip> module swap rocm rocm/6.0
sh dopamine hip> hipcc --version ; hipcc -o complex complex.hip.cc 
HIP version: 6.0.32831-204d35d16
AMD clang version 17.0.0 (https://github.com/RadeonOpenCompute/llvm-project roc-6.0.2 24012 af27734ed982b52a9f1be0f035ac91726fc697e4)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /apps/rocm/rocm-6.0.2/llvm/bin
Configuration file: /apps/rocm/rocm-6.0.2/lib/llvm/bin/clang++.cfg
complex.hip.cc:10:7: error: use of overloaded operator '*=' is ambiguous (with operand types 'hipFloatComplex' (aka 'HIP_vector_type<float, 2>') and 'hipFloatComplex')
    x *= y;
    ~ ^  ~
/apps/rocm/rocm-6.0.2/include/hip/amd_detail/amd_hip_vector_types.h:542:26: note: candidate function
        HIP_vector_type& operator*=(const HIP_vector_type& x) noexcept
                         ^
/apps/rocm/rocm-6.0.2/include/hip/amd_detail/amd_hip_complex.h:243:1: note: candidate function
COMPLEX_MUL_PREOP_OVERLOAD(hipFloatComplex)
^
/apps/rocm/rocm-6.0.2/include/hip/amd_detail/amd_hip_complex.h:112:41: note: expanded from macro 'COMPLEX_MUL_PREOP_OVERLOAD'
    __HOST_DEVICE__ static inline type& operator*=(type& lhs, const type& rhs) {                    \
                                        ^
complex.hip.cc:31:11: error: use of overloaded operator '*' is ambiguous (with operand types 'hipFloatComplex' (aka 'HIP_vector_type<float, 2>') and 'hipFloatComplex')
    x = x * y;
        ~ ^ ~
/apps/rocm/rocm-6.0.2/include/hip/amd_detail/amd_hip_vector_types.h:548:65: note: candidate function
        friend __HOST_DEVICE__ inline constexpr HIP_vector_type operator*(
                                                                ^
/apps/rocm/rocm-6.0.2/include/hip/amd_detail/amd_hip_complex.h:239:1: note: candidate function
COMPLEX_MUL_OP_OVERLOAD(hipFloatComplex)
^
/apps/rocm/rocm-6.0.2/include/hip/amd_detail/amd_hip_complex.h:80:40: note: expanded from macro 'COMPLEX_MUL_OP_OVERLOAD'
    __HOST_DEVICE__ static inline type operator*(const type& lhs, const type& rhs) {           \
                                       ^
complex.hip.cc:43:7: error: use of overloaded operator '*=' is ambiguous (with operand types 'hipDoubleComplex' (aka 'HIP_vector_type<double, 2>') and 'hipDoubleComplex')
    x *= y;
    ~ ^  ~
/apps/rocm/rocm-6.0.2/include/hip/amd_detail/amd_hip_vector_types.h:542:26: note: candidate function
        HIP_vector_type& operator*=(const HIP_vector_type& x) noexcept
                         ^
/apps/rocm/rocm-6.0.2/include/hip/amd_detail/amd_hip_complex.h:265:1: note: candidate function
COMPLEX_MUL_PREOP_OVERLOAD(hipDoubleComplex)
^
/apps/rocm/rocm-6.0.2/include/hip/amd_detail/amd_hip_complex.h:112:41: note: expanded from macro 'COMPLEX_MUL_PREOP_OVERLOAD'
    __HOST_DEVICE__ static inline type& operator*=(type& lhs, const type& rhs) {                    \
                                        ^
complex.hip.cc:44:11: error: use of overloaded operator '*' is ambiguous (with operand types 'hipDoubleComplex' (aka 'HIP_vector_type<double, 2>') and 'hipDoubleComplex')
    x = x * y;
        ~ ^ ~
/apps/rocm/rocm-6.0.2/include/hip/amd_detail/amd_hip_vector_types.h:548:65: note: candidate function
        friend __HOST_DEVICE__ inline constexpr HIP_vector_type operator*(
                                                                ^
/apps/rocm/rocm-6.0.2/include/hip/amd_detail/amd_hip_complex.h:261:1: note: candidate function
COMPLEX_MUL_OP_OVERLOAD(hipDoubleComplex)
^
/apps/rocm/rocm-6.0.2/include/hip/amd_detail/amd_hip_complex.h:80:40: note: expanded from macro 'COMPLEX_MUL_OP_OVERLOAD'
    __HOST_DEVICE__ static inline type operator*(const type& lhs, const type& rhs) {           \
                                       ^
4 errors generated when compiling for gfx90a.
```

It seems like perhaps it is resolved in 6.1rc1:
```
sh dopamine hip> module swap rocm rocm/6.1rc1
sh dopamine hip> hipcc --version ; hipcc -o complex complex.hip.cc 
HIP version: 6.1.40091-2d5941f89
AMD clang version 17.0.0 (https://github.com/RadeonOpenCompute/llvm-project roc-6.1.0 24084 1bd16cfd447ed0ff37fa4877e352f954c1088aa0)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /apps/rocm/rocm-6.1rc1/llvm/bin
Configuration file: /apps/rocm/rocm-6.1rc1/lib/llvm/bin/clang++.cfg
sh dopamine hip> 
```
Needs a little more verification.

**Comment by harkgill-amd** [comment]
Hi @mgates3, this issue is fixed as of ROCm 6.1.1. Thank you!
